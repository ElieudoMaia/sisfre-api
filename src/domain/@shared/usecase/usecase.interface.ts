export interface UseCase<P = void, R = void> {
  execute(params: P): Promise<R>;
}
