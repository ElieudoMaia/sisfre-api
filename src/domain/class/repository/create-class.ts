import { Class } from '../entity/class';

export interface CreateClassRepository {
  create(schoolClass: Class): Promise<void>;
}
