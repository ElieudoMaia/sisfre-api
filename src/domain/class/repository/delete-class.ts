export interface DeleteClassRepository {
  delete(id: string): Promise<void>;
}
