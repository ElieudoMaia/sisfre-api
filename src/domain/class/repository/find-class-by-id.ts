import { Class } from '../entity/class';

export interface FindClassByIdRepository {
  findById(id: string): Promise<Class | null>;
}
