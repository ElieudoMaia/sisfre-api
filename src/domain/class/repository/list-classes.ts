import { Class } from '../entity/class';

export type ListClassesRepositoryInput = {
  pageNumber?: number;
  pageSize?: number;
  description?: string;
};

export type ListClassesRepositoryOutput = {
  classes: Class[];
  quantity: number;
};

export interface ListClassesRepository {
  findAll: (
    params: ListClassesRepositoryInput
  ) => Promise<ListClassesRepositoryOutput>;
}
