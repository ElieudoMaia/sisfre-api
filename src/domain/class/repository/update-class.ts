import { Class } from '../entity/class';

export interface UpdateClassRepository {
  update(schoolClass: Class): Promise<void>;
}
