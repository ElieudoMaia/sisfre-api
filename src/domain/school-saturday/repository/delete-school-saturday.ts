export interface DeleteSchoolSaturdayRepository {
  delete: (schoolSaturdayId: string) => Promise<void>;
}
