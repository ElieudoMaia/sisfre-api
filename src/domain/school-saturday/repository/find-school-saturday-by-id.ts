import { SchoolSaturday } from '../entity/school-saturday';

export interface FindSchoolSaturdayByIdRepository {
  findById(id: string): Promise<SchoolSaturday | null | undefined>;
}
