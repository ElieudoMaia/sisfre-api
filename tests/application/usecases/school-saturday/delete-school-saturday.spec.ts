import { DeleteSchoolSaturdayUseCase } from '@/application/usecases/school-saturday/delete/delete-school-saturday.usecase';
import { DeleteSchoolSaturdayRepository } from '@/domain/school-saturday/repository/delete-school-saturday';
import { FindSchoolSaturdayByIdRepository } from '@/domain/school-saturday/repository/find-school-saturday-by-id';
import { describe, expect, it, vi } from 'vitest';
import { makeFakeSchoolSaturday } from '../../../@shared/fakes';
import { fake } from '../../../utils/fake-data-generator';

const schoolSaturdayId = fake.uuid();

const makeFindSchoolSaturdayByIdRepositoryStub =
  (): FindSchoolSaturdayByIdRepository => ({
    findById: vi
      .fn()
      .mockResolvedValue(makeFakeSchoolSaturday({ id: schoolSaturdayId }))
  });

const makeDeleteSchoolSaturdayRepositoryStub =
  (): DeleteSchoolSaturdayRepository => ({
    delete: vi.fn()
  });

type SutTypes = {
  sut: DeleteSchoolSaturdayUseCase;
  findSchoolSaturdayByIdRepositoryMock: FindSchoolSaturdayByIdRepository;
  deleteSchoolSaturdayRepositoryMock: DeleteSchoolSaturdayRepository;
};

const makeSut = (): SutTypes => {
  const findSchoolSaturdayByIdRepositoryMock =
    makeFindSchoolSaturdayByIdRepositoryStub();
  const deleteSchoolSaturdayRepositoryMock =
    makeDeleteSchoolSaturdayRepositoryStub();
  const sut = new DeleteSchoolSaturdayUseCase(
    findSchoolSaturdayByIdRepositoryMock,
    deleteSchoolSaturdayRepositoryMock
  );

  return {
    sut,
    findSchoolSaturdayByIdRepositoryMock,
    deleteSchoolSaturdayRepositoryMock
  };
};

describe('DeleteSchoolSaturdayUseCase', () => {
  it('should call FindSchoolSaturdayByIdRepository with correct params', async () => {
    const { sut, findSchoolSaturdayByIdRepositoryMock } = makeSut();
    await sut.execute({ id: schoolSaturdayId });
    expect(findSchoolSaturdayByIdRepositoryMock.findById).toHaveBeenCalledWith(
      schoolSaturdayId
    );
  });

  it('should throw if FindSchoolSaturdayByIdRepository throws', async () => {
    const { sut, findSchoolSaturdayByIdRepositoryMock } = makeSut();
    vi.spyOn(
      findSchoolSaturdayByIdRepositoryMock,
      'findById'
    ).mockRejectedValue(new Error('any_repo_error'));
    await expect(sut.execute({ id: schoolSaturdayId })).rejects.toThrow(
      'any_repo_error'
    );
  });

  it('should throw if FindSchoolSaturdayByIdRepository returns null', async () => {
    const { sut, findSchoolSaturdayByIdRepositoryMock } = makeSut();
    vi.spyOn(
      findSchoolSaturdayByIdRepositoryMock,
      'findById'
    ).mockResolvedValue(null);
    await expect(sut.execute({ id: schoolSaturdayId })).rejects.toThrow(
      `School Saturday with id ${schoolSaturdayId} not found`
    );
  });

  it('should call DeleteSchoolSaturdayRepository with correct params', async () => {
    const { sut, deleteSchoolSaturdayRepositoryMock } = makeSut();
    await sut.execute({ id: schoolSaturdayId });
    expect(deleteSchoolSaturdayRepositoryMock.delete).toHaveBeenCalledWith(
      schoolSaturdayId
    );
  });

  it('should throw if DeleteSchoolSaturdayRepository throws', async () => {
    const { sut, deleteSchoolSaturdayRepositoryMock } = makeSut();
    vi.spyOn(deleteSchoolSaturdayRepositoryMock, 'delete').mockRejectedValue(
      new Error('any_repo_error')
    );
    await expect(sut.execute({ id: schoolSaturdayId })).rejects.toThrow(
      'any_repo_error'
    );
  });

  it('should return undefined if everything succeeds', async () => {
    const { sut } = makeSut();
    const result = await sut.execute({ id: schoolSaturdayId });
    expect(result).toBeUndefined();
  });
});
