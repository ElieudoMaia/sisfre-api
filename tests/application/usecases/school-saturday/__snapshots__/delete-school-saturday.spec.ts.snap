// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DeleteSchoolSaturdayUseCase > should throw if FindSchoolSaturdayByIdRepository throws 1`] = `""`;
